utxo/src/utxo_impl/test.rs

```diff
@@ -0,0 +1,486 @@
+use common::primitives::{BlockHeight, Id, Idable, H256};
+
+use crate::utxo_impl::test_helper::Presence::{Absent, Present, Spent};
+use crate::Error::{self, OverwritingUtxo, UtxoAlreadyExists};
+use crate::{ConsumedUtxoCache, FlushableUtxoView, Utxo, UtxoEntry, UtxosCache, UtxosView};
+
+use crate::test_helper::create_tx_outputs;
+use crate::utxo_impl::test_helper::{
+    check_flags, create_utxo, create_utxo_for_mempool, insert_single_entry, Presence, DIRTY, FRESH,
+};
+use crate::utxo_impl::{UtxoSource, UtxoStatus};
+use common::chain::{OutPoint, OutPointSourceId, Transaction, TxInput};
+use crypto::random::{make_pseudo_rng, seq};
+use iter_tools::Itertools;
+use std::collections::BTreeMap;
+
+/// Checks `add_utxo` method behaviour.
+/// # Arguments
+/// `cache_presence` - initial state of the cache
+/// `cache_flags` - The flags of the existing utxo entry for testing
+/// `possible_overwrite` - to set the `possible_overwrite` of the `add_utxo` method
+/// `result_flags` - the result ( dirty/not, fresh/not ) after calling the `add_utxo` method.
+/// `op_result` - the result of calling `add_utxo` method, whether it succeeded or not.
+fn check_add_utxo(
+    cache_presence: Presence,
+    cache_flags: Option<u8>,
+    possible_overwrite: bool,
+    result_flags: Option<u8>,
+    op_result: Result<(), Error>,
+) {
+    let mut cache = UtxosCache::default();
+    let (_, outpoint) = insert_single_entry(&mut cache, &cache_presence, cache_flags, None);
+
+    // perform the add_utxo.
+    let (utxo, _) = create_utxo(0);
+    let add_result = cache.add_utxo(utxo, &outpoint, possible_overwrite);
+
+    assert_eq!(add_result, op_result);
+
+    if add_result.is_ok() {
+        let key = &outpoint;
+        let ret_value = cache.utxos.get(key);
+
+        check_flags(ret_value, result_flags, false);
+    }
+
+    println!("the value of the map: {:?}", cache);
+}
+
+/// Checks `spend_utxo` method behaviour.
+/// # Arguments
+/// `parent_presence` - initial state of the parent cache.
+/// `cache_presence` - initial state of the cache.
+/// `cache_flags` - The flags of a utxo entry in a cache.
+/// `result_flags` - the result ( dirty/not, fresh/not ) after performing `spend_utxo`.
+fn check_spend_utxo(
+    parent_presence: Presence,
+    cache_presence: Presence,
+    cache_flags: Option<u8>,
+    spend_result: Result<(), Error>,
+    result_flags: Option<u8>,
+) {
+    // initialize the parent cache.
+    let mut parent = UtxosCache::default();
+    let (_, parent_outpoint) =
+        insert_single_entry(&mut parent, &parent_presence, Some(FRESH | DIRTY), None);
```

Comment:
I think this `Some(FRESH | DIRTY)` is much nicer way of expressing the attributes

Reply:
Are you suggesting a specific change? I'm not sure I understand the request or whether it's just a comment with no implicit request.

Reply:
This was just a general comment referencing my earlier complaint about how IMO the booleans make the code unnecessarily hard to understand.